# TensorFlow Assignment: Convolutional Neural Network (CNN)

**[Duke Community Standard](http://integrity.duke.edu/standard.html): By typing your name below, you are certifying that you have adhered to the Duke Community Standard in completing this assignment.**

Name: Pratyush Sharma

### Convolutional Neural Network

Build a 2-layer CNN for MNIST digit classfication. Feel free to play around with the model architecture and see how the training time/performance changes, but to begin, try the following:

Image -> convolution (32 5x5 filters) -> nonlinearity (ReLU) ->  (2x2 max pool) -> convolution (64 5x5 filters) -> nonlinearity (ReLU) -> (2x2 max pool) -> fully connected (256 hidden units) -> nonlinearity (ReLU) -> fully connected (10 hidden units) -> softmax

Some tips:
- The CNN model might take a while to train. Depending on your machine, you might expect this to take up to half an hour. If you see your validation performance start to plateau, you can kill the training.

- Since CNNs a more complex than the logistic regression and MLP models you've worked with before, so you may find it helpful to use a more advanced optimizer. You're model will train faster if you use [`tf.train.AdamOptimizer`](https://www.tensorflow.org/api_docs/python/tf/train/AdamOptimizer) instead of `tf.train.GradientDescentOptimizer`. A learning rate of 1e-4 is a good starting point.

In [1]:
### YOUR CODE HERE ###
import tensorflow as tf

from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout, Activation, Flatten, Conv2D, MaxPooling2D

import matplotlib.pyplot as plt
from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical

#load data
(X_train, y_train), (X_test, y_test) = mnist.load_data()

#reshape data to fit model
X_train = X_train.reshape(60000,28,28,1)
X_test = X_test.reshape(10000,28,28,1)

#one-hot encode target column
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

#create model
model = Sequential()

#add model layers
model.add(Conv2D(32, kernel_size=5, activation='relu', input_shape=(28,28,1)))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Conv2D(64, kernel_size=5, activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Flatten())
model.add(Dense(256, activation='relu'))
model.add(Dense(10, activation='softmax'))

#compile model using accuracy as a measure of model performance
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])


#train model
model.fit(X_train, y_train,validation_data=(X_test, y_test), epochs=3)

#show predictions for the first 3 images in the test set
#model.predict(X_test[:4])

#show actual results for the first 3 images in the test set
#y_test[:4]



11493376/11490434 [==============================] - 0s 0us/step
Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor
Train on 60000 samples, validate on 10000 samples
Epoch 1/3
60000/60000 [==============================] - 225s 4ms/sample - loss: 0.3027 - acc: 0.9519 - val_loss: 0.0662 - val_acc: 0.9789
Epoch 2/3
60000/60000 [==============================] - 231s 4ms/sample - loss: 0.0628 - acc: 0.9808 - val_loss: 0.0648 - val_acc: 0.9804
Epoch 3/3
60000/60000 [==============================] - 229s 4ms/sample - loss: 0.0571 - acc: 0.9827 - val_loss: 0.0581 - val_acc: 0.9839


### Short answer

1\. How does the CNN compare in accuracy with yesterday's logistic regression and MLP models? How about training time?

After completing the three assignments, we get the following accuracy scores -
1) Logistic regression model - 91.85%
2) MLP model - 96.60%
3) CNN model - 98.27%
Clearly, CNN model accuracy is better than logistic regression and MLP models. 
But the training time is more than other models for CNN model. Where CNN takes approx. 15min to train, MLP and logistic regression models hardly took 2-3 mins.

2\. How many trainable parameters are there in the CNN you built for this assignment?

*Note: By trainable parameters, I mean individual scalars. For example, a weight matrix that is 10x5 has 50.*

 Conv1      Conv2        Dense1       Dense2 
(5x5x32) + (5x5x64) + (7x7x64x256) + (256x10) = 807,776

3\. When would you use a CNN versus a logistic regression model or an MLP?

The benefit of using CNNs is their ability to develop an internal representation of a two-dimensional image. Thus, CNN shouldbe used with Image data. CNNs work well with data that has a spatial relationship. They provide higher accuracy, at the cost of higher computation and training time.